psd/psd.c

```diff
@@ -111,68 +127,138 @@ int psd_dcdWrite(sdp_cmd_t *cmd)
 }
 
 
+int psd_getAttr(int type, offs_t* val, FILE* file)
+{
+	int i = 0;
+	msg_t msg;
+
+	while (i < FILES_SIZE && file != psd.files[i])
+		++i;
+
+	if (i == FILES_SIZE)
+		return -1;
+
+	msg.type = mtGetAttr;
+	msg.i.data = NULL;
+	msg.i.size = 0;
+	msg.o.data = NULL;
+	msg.o.size = 0;
+
+	msg.i.attr.type = type;
+	msg.i.attr.oid = psd.oids[i];
+
+	if (msgSend(psd.oids[i].port, &msg) < 0)
+		return -1;
+
+	*val = msg.o.attr.val;
+
+	return 0;
+}
+
+
 int psd_writeFile(sdp_cmd_t *cmd)
 {
-	int res, err = 0;
-	offs_t offs, n, l;
-	char buff[1025];
+	int res, err = 0, buffOffset = 0;
+	offs_t offs, flashoffs, partsz;
+
+	char *buff;
+	char *rcvBuff;
 	char *outdata;
 
-	if (fseek(psd.f, cmd->address, SEEK_SET) < 0)
-		err = -2;
+	flashoffs = cmd->address * FLASH_PAGE_SIZE;
 
-	for (offs = 0, n = 0; !err && (offs < cmd->datasz); offs += n) {
+	if (psd_getAttr(atSize, &partsz, psd.f) < 0)
+		err = -1;
 
-		/* Read data from serial device */
-		if ((res = psd.rf(1, buff, sizeof(buff), &outdata) < 0)) {
-			err = -1;
-			break;
+	if (!err && (cmd->datasz + flashoffs) > partsz)
+		err = -1;
+
+	if (!err && (fseek(psd.f, flashoffs, SEEK_SET) < 0))
+		err = -1;
+
+	rcvBuff = (char*) malloc(sizeof(char) * HID_DATA_SIZE);
```

Comment:
You should check if allocation was successful. Consider moving this buffers to .bss.